### PR TITLE
exec: implement int64 sorted distinct

### PR DESCRIPTION
**Pull request by jordanlewis**
This operator runs sorted distinct over n columns, populating the
selection vector with the indices of the tuples that are distinct in the
input.

This will be execgen templated later.

```
BenchmarkSortedDistinct-8         500000              3924 ns/op        6262.34 MB/s
```

Few other changes here:

- use []bool, not Bools, for boolean columns. Turns out the Bools idea was really slow - 3-4x worse on the benchmark.
- added a testing util library that uses a little Go reflection to make writing testcases easier.
- added the boolean zero operator

All based on top of #31554.

### CODE REVIEW

**Review by solongordon — CHANGES_REQUESTED** [review]
Could you add unit tests with a selection? Some of the logic looks suspicious to me.

> Reviewed 9 of 9 files at r1, 3 of 3 files at r2, 6 of 6 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-:-LPSYRcbDeU0hgpGEFzC:bjm48en)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/distinct.go, line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSV6tc4BryDmt9ijvl:-LPSV6tdEO_c977om_Zt:b-vyvpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L32)):*
> ```Go
> 
> 	// Starts at 1, flips to 0.
> 	firstColToLookAt uint16
> ```

I see what you did here, but I think a bool would make the intention clearer, even if it makes your code slightly more verbose. Also why `firstCol`? Isn't it more like the first row? The name is confusing me.
___
*[pkg/sql/exec/distinct.go, line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSRiMDBLFVu5oMLA6E:-LPSRiMDBLFVu5oMLA6F:bsxzu0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L52)):*
> ```Go
> 	outputCol := batch.ColVec(p.outputColIdx).Bool()
> 
> 	col := batch.ColVec(p.sortedDistinctCol).Int64()[:ColBatchSize]
> ```

What is the purpose of the `[:ColBatchSize]` slice?
___
*[pkg/sql/exec/distinct.go, line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSUBaR0t4IC2FwTwZc:-LPSUBaR0t4IC2FwTwZd:b-kriz3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L56)):*
> ```Go
> 	// we always output the first row.
> 	for i := uint16(0); i < p.firstColToLookAt; i++ {
> 		p.lastVal = col[0]
> ```

Shouldn't we take the selection vector into account here? What if the first row was filtered out?
___
*[pkg/sql/exec/distinct.go, line 108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSYE5dClXeZNe7osZ9:-LPSYE5dClXeZNe7osZA:ba1d8ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L108)):*
> ```Go
> 		max := sel[n-1]
> 		for i := uint16(0); i < max; i++ {
> 			if outputCol[i] {
> ```

This looks wrong to me. Don't we want `outputCol[sel[i]]`?

**Review by jordanlewis — COMMENTED** [review]
Thanks for the review. The logic was definitely suspicious / wrong. Will fix.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-:-LPS_9vs2F3tDT3n5mGz:b2fzlvn)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/distinct.go, line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSV6tc4BryDmt9ijvl:-LPSZtmICQzOzsDuT0wC:b-zeu6mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L32)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I see what you did here, but I think a bool would make the intention clearer, even if it makes your code slightly more verbose. Also why `firstCol`? Isn't it more like the first row? The name is confusing me.
</blockquote></details>

Ugh, yeah this makes no sense at all. Sorry. You're right on both accounts - I'll fix it.
___
*[pkg/sql/exec/distinct.go, line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSRiMDBLFVu5oMLA6E:-LPSZxmO0XPSHW5QL7iW:bw5o69s) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L52)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

What is the purpose of the `[:ColBatchSize]` slice?
</blockquote></details>

Prevents repeated bounds checks. It actually makes a difference in profiles.
___
*[pkg/sql/exec/distinct.go, line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSUBaR0t4IC2FwTwZc:-LPS_36D5OFwXE-emulM:b-y0rsr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L56)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Shouldn't we take the selection vector into account here? What if the first row was filtered out?
</blockquote></details>

:thumbsup: great point! I'll fix this.
___
*[pkg/sql/exec/distinct.go, line 108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSYE5dClXeZNe7osZ9:-LPS_8MPAWIq1SchhcNw:blrlsx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L108)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

This looks wrong to me. Don't we want `outputCol[sel[i]]`?
</blockquote></details>

Yes, this is fully broken.

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31693)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/distinct.go, line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSRiMDBLFVu5oMLA6E:-LPS_aAfCfuis4PJ9ipm:b-8dlil1) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L52)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Prevents repeated bounds checks. It actually makes a difference in profiles.
</blockquote></details>

Aha, good to know!

**Review by changangela — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31693)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/distinct.go, line 124 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSb0hg5b5zbQwLgFOY:-LPSb0hg5b5zbQwLgFOZ:btjlo96) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L124)):*
> ```Go
> 	}
> 
> 	batch.SetLength(idx)
> ```

if `idx==0`, will need to call `Next()` again.

**Review by jordanlewis — COMMENTED** [review]
Added randomized selection tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-:-LPWpVJYEYzbJsXIMKk5:bgxd0zt)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/distinct.go, line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSV6tc4BryDmt9ijvl:-LPWpUfW29wRODPm77HI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L32)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Ugh, yeah this makes no sense at all. Sorry. You're right on both accounts - I'll fix it.
</blockquote></details>

Done.
___
*[pkg/sql/exec/distinct.go, line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSUBaR0t4IC2FwTwZc:-LPWpTcq2miAzlLQfe2N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L56)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

:thumbsup: great point! I'll fix this.
</blockquote></details>

Done.
___
*[pkg/sql/exec/distinct.go, line 108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSYE5dClXeZNe7osZ9:-LPWpSpz-2JzkE9KWPEw:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L108)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Yes, this is fully broken. 
</blockquote></details>

Fixed.
___
*[pkg/sql/exec/distinct.go, line 124 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-LPSb0hg5b5zbQwLgFOY:-LPWpSRcDdKmEgcugRaQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58320c5723646bb4d15c1bb81a427157647c59da/pkg/sql/exec/distinct.go#L124)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

if `idx==0`, will need to call `Next()` again.
</blockquote></details>

Done.

File: pkg/sql/exec/distinct.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	startIdx := uint16(0)
+	if !p.foundFirstRow {
+		startIdx = 1
+	}
+
+	n := batch.Length()
+	if sel != nil {
+		// Bounds check elimination.
+		sel = sel[startIdx:n]
+		for s := 0; s < len(sel); s++ {
```
Comment (RaduBerinde) [inline pkg/sql/exec/distinct.go]: [nit] isn't `for s := range sel` the same?

File: pkg/sql/exec/distinct.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+			unique := col[i] != lastVal
+			outputCol[i] = outputCol[i] || unique
+			lastVal = col[i]
+		}
+	} else {
+		// Bounds check elimination.
+		col = col[startIdx:n]
+		outputCol = outputCol[startIdx:n]
+		for i := 0; i < len(col); i++ {
+			unique := col[i] != lastVal
```
Comment (RaduBerinde) [inline pkg/sql/exec/distinct.go]: [nit] I would add a comment here saying that it's important that `unique` is defined separately..
Or maybe rewrite as
```go
v := col[i]
outputCol[i] = outputCol[i] || (v != lastVal)
lastVal = v
```

File: pkg/sql/exec/distinct.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package exec
+
+// sortedDistinctInt64Op runs a distinct on the column in sortedDistinctCol,
+// writing the result to the bool column in outputColIdx by or'ing the
```
Comment (RaduBerinde) [inline pkg/sql/exec/distinct.go]: [nit] This comment needs some rephrasing (it sounds as if it's doing some bitwise OR stuff). I'd just say that the bool column is set to true for every value that differs from the previous one.

File: pkg/sql/exec/distinct.go (hunk 1 above)
Comment (jordanlewis) [reply pkg/sql/exec/distinct.go]: Yeah, missed that when doing the bounds check rewrite.

**Review by solongordon — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31693#-:-LPb6ttSAJVftBVwPY54:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by jordanlewis** [comment]
TFTRs!

bors r+

**Pull request merged**